### PR TITLE
[Keyboard] Add YDKB Mountain

### PR DESCRIPTION
**Pull request by Wraul**
This keyboard comes with a YDKB firmware that can be configured using the tool at https://ydkb.io/
But I want to use QMK with all my keyboards, so here we are.

## Description

Mapping the PCB matrix was a bit of a pain as it uses multiple 74HC595 shift registers, which I never worked with before. But after some reverse enginering and looking at the YDKB firmware I have something that works. Please let me know if there is something that could be done better.

I used the same USB PID and VID as the YDKB firmware. This might be a good thing or maybe I shouldn't do that?

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by sigprof** [comment]
Great — I always wondered how they managed to make a keyboard with what looks like shift register chips, but without any resistors or diodes.

Can you share the exact markings of the “74HC595” chips that are installed on the PCB?  Apparently they are not the usual 74HC595 variety with tri-state push-pull outputs, but a special variation with open drain outputs.  In https://electronics.stackexchange.com/questions/512675/probable-issue-with-74hc595-not-outputting-any-data someone was able to find an example of such chip (https://www.lcsc.com/product-detail/74-Series_Shenzhen-Fuman-Elec-74HC595A_C110383.html); the [datasheet](https://datasheet.lcsc.com/lcsc/2202252130_Shenzhen-Fuman-Elec-74HC595A_C110383.pdf) is in Chinese, and the Google translation looks like “The 74HC595A storage register has 8 bus driver outputs, and the data output mode is low level and high impedance”, which apparently means open drain outputs.

File: keyboards/ydkb/mountain/readme.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+The following two pins are used to read the state of the matrix.
+
+| ATmega Pin | Function   |
+| ---------- | ---------- |
+| `B2`       | Right half |
+| `B3`       | Left half  |
+
+As shown above the pin `B3` is used both to control the 74HC595 and to read the
+state of the matrix. Thus it has to be switch between output and input modes
+between the two phases.
```
Comment (sigprof) [inline keyboards/ydkb/mountain/readme.md:77]: I still don't fully understand how this is supposed to work at the schematic level (when some key is pressed, the `B3` pin would be pulled low by the shift register output through the switch when that key is scanned, but at the next step `B3` needs to be forced high in order to write 1 into the shift register).  Or are the SER and key status signals actually separate at the connector, and there is some resistor on the controller board to allow the MCU pin output to override the input from the key?

**Review by sigprof — CHANGES_REQUESTED** [review]

**Comment by Wraul** [comment]
> Can you share the exact markings of the “74HC595” chips that are installed on the PCB? Apparently they are not the usual 74HC595 variety with tri-state push-pull outputs, but a special variation with open drain outputs. In https://electronics.stackexchange.com/questions/512675/probable-issue-with-74hc595-not-outputting-any-data someone was able to find an example of such chip (https://www.lcsc.com/product-detail/74-Series_Shenzhen-Fuman-Elec-74HC595A_C110383.html); the [datasheet](https://datasheet.lcsc.com/lcsc/2202252130_Shenzhen-Fuman-Elec-74HC595A_C110383.pdf) is in Chinese, and the Google translation looks like “The 74HC595A storage register has 8 bus driver outputs, and the data output mode is low level and high impedance”, which apparently means open drain outputs.

OMG thank you!
This explains why I wasn't able to get any output on the register pins during my experiments while trying to figure out how the thing works. Turns out I'm not crazy after all.

I have added the markings to the readme.md
There is also a SM Micro logo. Like the one below, but without the text.
logo

I can try to get you a photo if you like. Just let me know.

File: keyboards/ydkb/mountain/info.json [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    "matrix_size": {
+        "cols": 16,
+        "rows": 5
+    },
+    "processor": "atmega32u4",
+    "url": "https://kbdfans.com/products/ic-kbdfans-mountain-ergo-keyboard-kit",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x2163",
+        "vid": "0xFEED"
```
[the remark this answers is not in the document]
Comment (Wraul) [reply keyboards/ydkb/mountain/info.json]: I have switched the vendor ID to `0x5945` which appears to be used by a few other YDKB keyboards in QMK.

File: keyboards/ydkb/mountain/info.json [hunk 3]
```diff
@@ -0,0 +1,170 @@
+{
+    "manufacturer": "YDKB",
+    "keyboard_name": "Mountain Keyboard",
+    "maintainer": "Wraul",
+    "bootloader": "lufa-ms",
+    "features": {
+        "bootmagic": true,
+        "command": true,
```
[the remark this answers is not in the document]
Comment (Wraul) [reply keyboards/ydkb/mountain/info.json]: Sure thing.
I have disabled the command feature.

File: keyboards/ydkb/mountain/matrix.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 Mathias Andersson (@Wraul)
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "config.h"
```
[the remark this answers is not in the document]
Comment (Wraul) [reply keyboards/ydkb/mountain/matrix.c]: I think I copied the initial `matrix.c` from somewhere.
But the include should now be removed.

File: keyboards/ydkb/mountain/readme.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+### Flashing on Linux
+
+Flashing a firmware can be done on Linux by entering the bootloader and using
+`dd` as described below.
+
+The storage device can differ between systems, so make sure to specify the correct device.
+To list available devices `lsblk` can be used.
+
+```
+dd if=./ydkb_mountain_{LAYOUT}.bin of=/dev/sd{X} seek=4 && sync
```
[the remark this answers is not in the document]
Comment (Wraul) [reply keyboards/ydkb/mountain/readme.md:43]: I got the command from the YDKB documentation [here](https://ydkb.io/help/#/en/bootloader/msd-bootloader?id=reflash-firmware-in-linux).

Do you want me to update the documentation to use your command instead?

File: keyboards/ydkb/mountain/readme.md (hunk 1 above)
Comment (Wraul) [reply keyboards/ydkb/mountain/readme.md:87]: I don't think my electronics skills are enough to understand the exact details. But yes there is a resistor. If I remember correctly it is `R1` in the bottom right of [this picture](https://kbdfans.com/cdn/shop/products/6_02a8c17a-99dd-4540-9cfd-40642542aa0c.jpg?v=1649294337). I think it was in the mega ohm range. And it caused me some grief when I tried to do some continuity probing when figuring out how the keys on the right PCB was scanned.

If you want me to I can check the exact specs and try to figure out exactly how it is connected.

**Comment by Wraul** [comment]
> Thank you for your contribution!
> This pull request has been automatically marked as stale because it has not had activity in the last 45 days. It will be closed in 30 days if no further activity occurs. Please feel free to give a status update now, or re-open when it's ready.
> For maintainers: Please label with `bug`, `awaiting review`, `breaking_change`, `in progress`, or `on hold` to prevent the issue from being re-flagged.

Are I expected to do anything? Cause at this point I don't know of anything that needs to change with this PR

**Review by drashna — APPROVED** [review]

File: keyboards/ydkb/mountain/info.json [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        "mousekey": true,
+        "nkro": true
+    },
+    "indicators": {
+        "caps_lock": "D5"
+    },
+    "matrix_size": {
+        "cols": 16,
+        "rows": 5
+    },
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/info.json:20]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/ydkb/mountain/info.json (hunk 6 above)
Comment (Wraul) [reply keyboards/ydkb/mountain/info.json:20]: I don't understand. If I remove the "matrix_size" it won't compile?
Am I missing something?

```
make ydkb/mountain:default
QMK Firmware 0.23.9
☒ 'matrix_size'
Traceback (most recent call last):
  File "/nix/store/9h4bn3gg1qn5byjvbyffmv8q1p8289bq-python3-3.10.11-env/lib/python3.10/site-packages/milc/milc.py", line 528, in __call__
    return self.__call__()
  File "/nix/store/9h4bn3gg1qn5byjvbyffmv8q1p8289bq-python3-3.10.11-env/lib/python3.10/site-packages/milc/milc.py", line 533, in __call__
    return self._subcommand(self)
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/decorators.py", line 27, in wrapper
    return func(*args, **kwargs)
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/cli/list/layouts.py", line 21, in list_layouts
    info_data = info_json(cli.config.list_layouts.keyboard)
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/info.py", line 216, in info_json
    info_data = _extract_led_config(info_data, str(keyboard))
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/info.py", line 706, in _extract_led_config
    cols = info_data['matrix_size']['cols']
KeyError: 'matrix_size'
Making ydkb/mountain with keymap default

☒ 'matrix_size'
Traceback (most recent call last):
  File "/nix/store/9h4bn3gg1qn5byjvbyffmv8q1p8289bq-python3-3.10.11-env/lib/python3.10/site-packages/milc/milc.py", line 528, in __call__
    return self.__call__()
  File "/nix/store/9h4bn3gg1qn5byjvbyffmv8q1p8289bq-python3-3.10.11-env/lib/python3.10/site-packages/milc/milc.py", line 533, in __call__
    return self._subcommand(self)
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/cli/generate/rules_mk.py", line 65, in generate_rules_mk
    kb_info_json = dotty(info_json(cli.args.keyboard))
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/info.py", line 216, in info_json
    info_data = _extract_led_config(info_data, str(keyboard))
  File "/home/mathias/Developer/qmk_firmware/lib/python/qmk/info.py", line 706, in _extract_led_config
    cols = info_data['matrix_size']['cols']
KeyError: 'matrix_size'

 *                                                                                                     [ERRORS]
```

Comment (fauxpark) [reply keyboards/ydkb/mountain/info.json:20]: `matrix_size` is not part of the `info.json` schema, though it is used internally. For custom matrix, the matrix dimensions need to be specified in `config.h`.

Comment (Wraul) [reply keyboards/ydkb/mountain/info.json]: I see. Thank you for the explanation.

This should hopefully be fixed now.

File: keyboards/ydkb/mountain/matrix.c [hunk 7]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Mathias Andersson (@Wraul)
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "quantum.h"
+
+void pulse_registers(void) {
+    writePinHigh(B1);
+    asm("nop");
+    writePinLow(B1);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:8]:
```suggestion
    gpio_write_pin_high(B1);
    asm("nop");
    gpio_write_pin_low(B1);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+void matrix_read_mode(void) {
+    setPinInputHigh(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:12]:
```suggestion
    gpio_set_pin_input_high(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    wait_us(5);
+}
+
+void matrix_write_mode(void) {
+    setPinOutput(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:17]:
```suggestion
    gpio_set_pin_output(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+void reset_registers(void) {
+    matrix_write_mode();
+    for (uint8_t i = 0; i < 40; i++) {
+        writePinHigh(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:23]:
```suggestion
        gpio_write_pin_high(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        pulse_registers();
+    }
+    writePinLow(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:26]:
```suggestion
    gpio_write_pin_low(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    pulse_registers();
+    matrix_read_mode();
+}
+
+void shift_registers(void) {
+    matrix_write_mode();
+    writePinHigh(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:33]:
```suggestion
    gpio_write_pin_high(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    matrix_read_mode();
+}
+
+void matrix_init_custom(void) {
+    setPinOutput(B1);
+    setPinInputHigh(B2);
+    setPinOutput(B3);
+
+    writePinHigh(B1);
+    writePinHigh(B3);
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:44]:
```suggestion
    gpio_set_pin_output(B1);
    gpio_set_pin_input_high(B2);
    gpio_set_pin_output(B3);

    gpio_write_pin_high(B1);
    gpio_write_pin_high(B3);
```

File: keyboards/ydkb/mountain/matrix.c [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    matrix_row_t current_row_value;
+
+    reset_registers();
+
+    for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
+        current_row_value = 0;
+
+        for (uint8_t current_col = 0; current_col < MATRIX_COLS / 2; current_col++) {
+            /* Left hand scan (Bank U1) */
+            if (!readPin(B3)) {
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:58]:
```suggestion
            if (!gpio_read_pin(B3)) {
```

File: keyboards/ydkb/mountain/matrix.c [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+                current_row_value |= (matrix_row_t)1 << current_col;
+            }
+            /* Right hand scan (Bank U2) */
+            if (!readPin(B2)) {
```
Comment (fauxpark) [inline keyboards/ydkb/mountain/matrix.c:62]:
```suggestion
            if (!gpio_read_pin(B2)) {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by Wraul** [comment]
I have now updated the code to use the new macro names.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/ydkb/mountain/matrix.c [hunk 16]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Mathias Andersson (@Wraul)
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "quantum.h"
```
[the remark this answers is not in the document]
Comment (Wraul) [reply keyboards/ydkb/mountain/matrix.c]: I have changed the include as instructed. But I also had to add `wait.h`, as the matrix code uses `wait_us()`. Hopefully that's the correct approach.

I also rebased the PR onto the latest master and renamed `info.json`  to `keyboard.json` as indicated by the make warning.

**Comment by Wraul** [comment]
I have once again rebased this on the latest master.

**Pull request closed without merging**